runtime/parser2/parser.go

```diff
@@ -229,13 +231,27 @@ func (p *parser) acceptBuffered() {
 	}
 }
 
+// replayLimit is a sensible limit for how many tokens may be replayed
+// while parsing a program
+const replayLimit = 2 << 12
```

Comment:
How did we come up with this value? 

It seems like this is setting a total limit on the amount of backtracking for the whole program. Would it not be better to instead impose a (smaller) limit on the amount of backtracking that can happen at once? I.e. impose a limit for each backtracking cursor on the stack instead of a total limit for the stack as a whole?

This has the benefit of being able to localize errors for users should they encounter them; instead of complaining that the entire program is too ambiguous at some arbitrary location where we pass the total limit, we can error at the actual location that has too much backtracking.

Reply:
I determined it mostly through experimentation, there is no perfect value. We need to find a limit that rejects invalid programs as quickly as possible, while not rejecting valid programs.

There are problems with both local and global limits. With the global limit, the user may encounter the limit in a locally small ambiguity, because the parser already re-parsed too many tokens before. With the local limit, finding a sensible limit is hard: A small value will reject too many programs, a slightly larger value will not reject invalid programs quickly.

Maybe we can have both local and global limits to balance user experience and security.

Reply:
Another idea: Maybe we can enforce the limit locally to each top-level ambiguity, instead of the whole program? We could reset `replayedCount` in `acceptBuffered`

Reply:
Made the token replay limit "local" to each top-level reparse in f2c81d9. Nested replays count towards the top-most buffering, so errors stay local.